src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineRetentionStrategy.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.jenkins.plugins.computeengine;
+
+import com.google.common.collect.ImmutableList;
+import hudson.model.Action;
+import hudson.model.Cause;
+import hudson.model.CauseAction;
+import hudson.model.Executor;
+import hudson.model.ExecutorListener;
+import hudson.model.Job;
+import hudson.model.Queue;
+import hudson.security.ACL;
+import hudson.security.ACLContext;
+import hudson.slaves.RetentionStrategy;
+import java.util.List;
+import java.util.logging.Level;
+import jenkins.model.Jenkins;
+import lombok.extern.java.Log;
+import org.jenkinsci.plugins.durabletask.executors.OnceRetentionStrategy;
+
+/**
+ * A strategy that allows: - setting one shot instances {@link OnceRetentionStrategy} - in case of
+ * preemption of GCP instance to restart preempted tasks
+ */
+@Log
+public class ComputeEngineRetentionStrategy extends RetentionStrategy<ComputeEngineComputer>
+    implements ExecutorListener {
+
+  private final OnceRetentionStrategy delegate;
+  private final boolean oneShot;
+
+  /**
+   * Creates the retention strategy.
+   *
+   * @param retentionTimeMinutes number of minutes of idleness after which to kill the slave; serves
+   *     a backup in case the strategy fails to detect the end of a task
+   * @param oneShot create one shot instance strategy
+   */
+  ComputeEngineRetentionStrategy(int retentionTimeMinutes, boolean oneShot) {
+    this.oneShot = oneShot;
+    delegate = new OnceRetentionStrategy(retentionTimeMinutes);
+  }
+
+  @Override
+  public long check(ComputeEngineComputer c) {
+    return delegate.check(c);
```

Comment:
These overrides should be calling their respective super implementations.

Reply:
Also, for check, do we want to be calling the OnceRetention's check if the instance isn't one-shot?

Reply:
> These overrides should be calling their respective super implementations.

They are calling them..
Not super but delegate..

Reply:
> Also, for check, do we want to be calling the OnceRetention's check if the instance isn't one-shot?

No.. we need to call it actually..

Reply:
Could you expand on why please.

Reply:
On `start` we actually need to connect machines..
And on `check` we need to call delegate to remove old slaves (non one shot).

Reply:
My concern with this approach, is OnceRetentionStrategy.check() will shutdown idling machines. This seems to be the opposite of what we want when oneshot is false.

Reply:
Yes.. thats exactly the flow that we have now.

Either machine is one shotted and should be closed in `taskCompleted*` or is not one shot and should be closed when idle by `check`.

Reply:
Previous flow `OnceRetentionStrategy` was for one show and `CloudRetentionStrategy` was for killing idle..
```
.retentionStrategy(
              oneShot
                  ? new OnceRetentionStrategy(retentionTimeMinutes)
                  : new CloudRetentionStrategy(retentionTimeMinutes))
```

Reply:
Understood. Thanks for clarifying.